File: preprocess/ts_filter.m
%% Setup
function ts_filter(sub_name, acq_type, n_session, cutoff_t)
addpath('cifti-matlab');

% sub_name: Subject label
% acq_type: Acquisition label
% n_session: Number of sessions

base_dir = strcat('~/Data/fMRI', '/', sub_name, '/', acq_type);

% default cutoff temporal frequency
if ~exist('cutoff_t','var')
    cutoff_t = 150.0;
end

%% Run preprocessing for all sessions
ses_idx = 1 : n_session;

for idx = ses_idx
    % load the data file for each session
    ses_name = sprintf('func-%02d', idx);
    [cifti_data, motion_rg] = load_data(base_dir, sub_name, ses_name);

    % cifti time series
    ts = cifti_data.cdata';

    % cutoff frequency
    spRate = 1 / 0.80;
    cutoff = 1 / cutoff_t;
    ts = highpass(ts, cutoff, spRate);

    % setup nuisance variables
    % decorrelation using PCA
    [~, score, latent] = pca(motion_rg);

    latent = latent / sum(latent);
    cum_lt = cumsum(latent);

    cutoff = sum(cum_lt < 1-1e-3);
    score = score(:, 1:cutoff);

    % setup motion nuisance regressors
    rgs = [score, ones(size(ts, 1), 1)];

    % solve with normal equation
    % save the residule as new time series
    theta = (rgs' * rgs) \ (rgs' * ts);
    ts = ts - rgs * theta;

    % z-score normalization
    meanVec = mean(ts, 1);
    stdVec = std(ts, 0, 1);
    ts = (ts - meanVec) ./ stdVec;

    % save output as icafix output name
    cifti_data.cdata = ts';
    data_file = fullfile(base_dir, strcat(ses_name, '_Atlas_hp2000_clean.dtseries.nii'));
    cifti_write(cifti_data, data_file);
end

end

File: preprocess/ts_filter.py
import os

# Setup variables
sub_name = 'HERO_LZ'
acq_type = 'pRF'
home = os.path.expanduser('~')
base = os.path.join(home, 'Data', 'fMRI',
                        sub_name, acq_type)
base_dir = base

# Create file hierarchy
dir_name = ['Filtered', sub_name, 'MNINonLinear', 'Results']
for dir in dir_name:
    base_dir = os.path.join(base_dir, dir)
    if not os.path.exists(base_dir):
        os.system("mkdir %s" % base_dir)

# Move data files to corresponding folders
# This step comes after the MATLAB script
n_session = 6
file_base = '_Atlas_hp2000_clean.dtseries.nii'
for idx in range(n_session):
    ses_name = 'func-0%d' % (idx + 1)
    ses_dir = os.path.join(base_dir, ses_name)

    if not os.path.exists(ses_dir):
        os.system("mkdir %s" % ses_dir)

    file_name = ses_name + file_base
    file_path = os.path.join(ses_dir, file_name)
    if not os.path.exists(file_path):
        source_path = os.path.join(base, file_name)
        os.system("mv %s %s" % (source_path, ses_dir))

# Create a zip file
os.chdir(os.path.join(base, 'Filtered'))
if not os.path.exists('%s.zip' % sub_name):
    os.system("zip -r %s %s" % (sub_name, sub_name))